fix: check contract balance before deposit

felixfaisal opened the pull request:
It checks if there's enough balance in the contract for a payout, else it exits before making a deposit.

felixfaisal committed:
fix: check contract balance before deposit

felixfaisal committed:
refactor: cargo fmt

felixfaisal committed:
refactor: cargo clippy

silva-fj commented on tee-worker/omni-executor/intent/executors/cross-chain/src/lib.rs:
```diff
@@ -603,6 +603,36 @@ impl<Provider: EthereumRpcProvider<Transaction = TransactionRequest> + Send + Sy
+						from_amount_decimal,
+					)
+					.await?;
+					let payout_amount = str_to_u256(&estimated_bnb_receive, 18).ok_or(())?;
```
We can also move the payout_amount conversion we do in the next steps here, so we also notify pumpx BE  and resolve the TODO

This part: 
```rust
let payout_amount = match str_to_u256(&bnb_to_receive, 18) {
	Some(a) => a,
	None => {
		log::error!("Fail to convert bnb amount {} to U256", bnb_to_receive);
		let body = CrossFailBody {
			request_id: intent_id,
			fail_reason:
			"Fail to construct payout request due to U256 conversion error"
				.to_string(),
		};
		self.pumpx_api.cross_fail(&access_token, body).await.map_err(|_| {
			log::error!("Failed to notify pumpx-signer");
		})?;
		// TODO: what to do with user asset?
		return Err(());
	},
};
```

felixfaisal replied:
I've updated the code, to inform `pumpx` BE, but `payout` calculation still need to take place again if we're waiting for the trade to be completed based on trade fill data, unless i'm missing something here. 

I do have one question though, what is the criteria for informing `pumpx` BE, I couldn't find a pattern here, some parts of the code we are simply logging the error and returning `Err()`, and in some we are informing `pumpx` BE.

silva-fj replied:
> I've updated the code, to inform pumpx BE, but payout calculation still need to take place again if we're waiting for the trade to be completed based on trade fill data, unless i'm missing something here.

Yes, you are right. Actually I found a couple of bugs while looking at this. I have opened a small PR to your branch to address them: https://github.com/litentry/heima/pull/3421

> I do have one question though, what is the criteria for informing pumpx BE, I couldn't find a pattern here, some parts of the code we are simply logging the error and returning Err(), and in some we are informing pumpx BE.

Yes we need to iterate on this, we have covered only the happy path

felixfaisal replied:
I've merged it, Let me know if there's something else i need to look at.

felixfaisal committed:
refactor: pr comments

silva-fj committed:
Fix bnb_to_receive and payout_amount (#3421)

* refactor: rename bnb_received to bnb_acquired

Renamed the variable `bnb_received` to `bnb_acquired` to better reflect its
purpose. This variable represents the amount of BNB resulting from a trade
executed via the Binance spot trading API, which is held in the service's
Binance account and not yet transferred to the user's account.

* fix(cross-chain): assign acquired BNB to bnb_to_receive

Set `bnb_to_receive` to `bnb_acquired` after successful acquisition,
ensuring the correct value is used in subsequent logic.

felixfaisal committed:
Merge branch 'dev' into p-1480-pre-check-the-remaining-bnb-in-the-accounting-contract
